### PR TITLE
[dynamo] Added support for type for generated custom object 

### LINKED ISSUE #109056: UserError: Can't call type() on generated custom object. Please use __class__ instead
### 🐛 Describe the bug

```
    File "/data/users/ezyang/b/pytorch/torch/_dynamo/symbolic_convert.py", line 1119, in CALL_FUNCTION                                 self.call_function(fn, args, {})                                                                                           
    File "/data/users/ezyang/b/pytorch/torch/_dynamo/symbolic_convert.py", line 565, in call_function                                  self.push(fn.call_function(self, args, kwargs))                                                                            
    File "/data/users/ezyang/b/pytorch/torch/_dynamo/variables/builtin.py", line 618, in call_function                                 result = handler(tx, *args, **kwargs)                                                                                      
    File "/data/users/ezyang/b/pytorch/torch/_dynamo/variables/builtin.py", line 1220, in call_type                              
      raise UserError(                                                                                                           
  torch._dynamo.exc.UserError: Can't call type() on generated custom object. Please use __class__ instead 
```

this error seems nonsensical to me, we should just support it

showed up in torchrec `dist_data.py`

```
        return type(self._input).dist_init(
            keys=self._keys,
            tensors=self._output_tensors,
            batch_size_per_rank=self._batch_size_per_rank,
            recat=self._recat,
        )
```

### Versions

main

cc @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by haixu-qin**
Fixes #109056

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @ezyang for the issue.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by ezyang** [comment]
Needs test. Also can you find the person who originally added that error and add them as reviewer

**Comment by haixu-qin** [comment]
Test added.

File: test/dynamo/test_functions.py [hunk 1]
```diff
@@ -470,7 +470,11 @@ def test_get_default_dtype(x):
     def test_get_autocast_gpu_dtype(x):
         dtype = torch.get_autocast_gpu_dtype()
         return x.type(dtype)
-
+      
+    @make_test
+    def test_generated_custom_object_types(x):
+        return type(x)
```
Comment (ezyang) [inline test/dynamo/test_functions.py:476]: x is a tensor, this seems pretty unlikely to trigger the error case...

Comment (haixu-qin) [reply test/dynamo/test_functions.py]: test edited. @ezyang .

Comment (haixu-qin) [reply test/dynamo/test_functions.py]: Could you review it? @ezyang .

**Comment by ezyang** [comment]
no this is still not right. You must return a VT but you are not. From there, I can infer your test is not working.

**Pull request closed without merging**